FeedStoreChallenge/CoreDataFeedStore/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,131 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Edgar Hirama on 08/03/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import CoreData
+
+public class CoreDataFeedStore: FeedStore {
+
+	enum CoreDataError: Error {
+		case unableToFindModel
+		case unableToLoad(Error)
+	}
+
+	private let storeURL: URL
+	private let bundle: Bundle
+
+	private static let resourceName: String = "FeedStore"
+
+	private let persistentContainer: NSPersistentContainer
+	private let context: NSManagedObjectContext
+
+	public init(storeURL: URL, bundle: Bundle = .main) throws {
+		self.storeURL = storeURL
+		self.bundle = bundle
+		self.persistentContainer = try CoreDataFeedStore.loadPersistendContainer(storeURL: storeURL, bundle: bundle)
+		self.context = persistentContainer.newBackgroundContext()
+	}
+
+	private static func loadPersistendContainer(storeURL: URL, bundle: Bundle) throws -> NSPersistentContainer {
+		guard let bundleURL = bundle.url(forResource: CoreDataFeedStore.resourceName, withExtension: "momd"),
+			  let model = NSManagedObjectModel(contentsOf: bundleURL) else { throw CoreDataError.unableToFindModel }
+
+		let container = NSPersistentContainer(name: CoreDataFeedStore.resourceName, managedObjectModel: model)
+		let description = NSPersistentStoreDescription(url: storeURL)
+		container.persistentStoreDescriptions = [description]
+
+		var loadError: Error?
+		container.loadPersistentStores { _, error in
+			loadError = error
+		}
+		if let error = loadError {
+			throw CoreDataError.unableToLoad(error)
+		}
+		return container
+	}
+
+	private func deleteCurrentCacheIfNeeded(in context: NSManagedObjectContext) {
+		let fetchRequest: NSFetchRequest<LocalCache> = NSFetchRequest(entityName: LocalCache.className())
+		if let fetchResult = try? context.fetch(fetchRequest), let existingFeed = fetchResult.first {
+			context.delete(existingFeed)
+		}
+	}
+}
+
+extension CoreDataFeedStore {
+
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		let managedContext = context
+		do {
+			let fetchRequest = NSFetchRequest<LocalCache>(entityName: "LocalCache")
+
+			let result = try managedContext.fetch(fetchRequest)
+			if let cacheResult = result.first {
+				completion(.found(feed: cacheResult.feed.array.compactMap { ($0 as? CacheFeedImage)?.localFeedImage } , timestamp: cacheResult.timestamp))
```

Comment:
You can create a `localFeed` helper to simplify this line here:

```suggestion
				completion(.found(feed: cacheResult.localFeed, timestamp: cacheResult.timestamp))
```